### DISCUSSION TITLE
Does Terrakube Support Uploading of Code

### DISCUSSION THREAD

**Discussion by JPatC** [terrakube-io/terrakube#1447] [Q&A] [answered]
We're migrating from Terraform Enterprise to Terrakube and our build process currently interacts with terraform enterprise to create a workspace version. This version provides an upload URL that we use to upload our code. Does terrakube provide any ability for us to upload (POST) our code?

We use the "Configuration Version" of the workspace and the Direct Upload where we can post a zipped file of our code:
https://developer.hashicorp.com/terraform/enterprise/workspaces/configurations

**Comment by alfespa17** [comment, marked as answer]
Hello @JPatC Terrakube support some of the TFC API methods (only the ones that are used for the CLI driven workflow).

I think these are the endpoints that you need:

to create the configuration version.

"POST /remote/tfe/v2/workspaces/{workspaceId}/configuration-versions"

https://github.com/AzBuilder/terrakube/blob/5639fec67c1a16d95fac56e0e499d01b8ca3a8a2/api/src/main/java/org/terrakube/api/plugin/state/RemoteTfeController.java#L210

to upload the tar.gz file using the response property upload-url from the above method:

"PUT /remote/tfe/v2/configuration-versions/{configurationid}"

https://github.com/AzBuilder/terrakube/blob/5639fec67c1a16d95fac56e0e499d01b8ca3a8a2/api/src/main/java/org/terrakube/api/plugin/state/RemoteTfeController.java#L227

Those 2 entpoints are the one used in the terraform code here when running a terraform remote plan

https://github.com/hashicorp/terraform/blob/2e4a62d92906399cf8e079e0513fde9df14b2537/internal/cloud/backend_plan.go#L140

https://github.com/hashicorp/terraform/blob/2e4a62d92906399cf8e079e0513fde9df14b2537/internal/cloud/backend_plan.go#L200

Once you have uploaded the source code using a "tar.gz" you will have to update the workspace "source" property, Terrakube does that when creating the job in this part of the code

https://github.com/AzBuilder/terrakube/blob/5639fec67c1a16d95fac56e0e499d01b8ca3a8a2/api/src/main/java/org/terrakube/api/plugin/state/RemoteTfeService.java#L798

https://github.com/AzBuilder/terrakube/blob/5639fec67c1a16d95fac56e0e499d01b8ca3a8a2/api/src/main/java/org/terrakube/api/plugin/state/RemoteTfeService.java#L803

**Reply by alfespa17** [reply]
By the way there is also an importer feature (I don't remember if that migrate the files too) 

https://docs.terrakube.io/user-guide/migrating-to-terrakube#migrating-with-the-workspaces-importer

**Reply by JPatC** [reply]
Thanks this is amazing! Is there documentation of these /remote/tfe/v2/* APIs? And can I only use these with the CLI driven workflow? I was reading up on that and it looks like we wouldn't be able to apply the plan in terrakube if we have to use the remote CLI flow which I found here https://docs.terrakube.io/user-guide/workspaces/cli-driven-workflow 

Do you know if the APIs support OpenAPI (swagger) so I can get the documentation?

**Reply by alfespa17** [reply]
Hello @JPatC you wont find any documentation for the "/remote/tfe/v2/" endpoints, but you could check this [file](https://github.com/AzBuilder/terrakube/blob/main/api/src/main/java/org/terrakube/api/plugin/state/RemoteTfeController.java) to see the available methods.

To create a plan using the TFC API I think you need to use the following method if I remembered correctly

https://github.com/AzBuilder/terrakube/blob/e30c9863d034fb8b7908ce566dab2a08a276215d/api/src/main/java/org/terrakube/api/plugin/state/RemoteTfeController.java#L237

You could use those endpoints if you want (as a mentioned above it only provide support for the methods and fileds required when you are using the "cloud block" and the "remote backend" used in the CLI driven workflow)

Terrakube provide its own API that is much simpler than the TFC API.

You can get the API documentation if you do a "get request" to the following endpoint "terrakube-api.mydomain.com/doc" as you can see [here](https://github.com/AzBuilder/terrakube/pull/1009).

For example to create a "plan" you can do this simple request

```shell
POST /api/v1/organization/${organizationId}/job
Content-Type: application/vnd.api+json
Authorization: PAT TOKEN
{
      data: {
        type: "job",
        attributes: {
          templateReference: ${templateId}, <-- This should be the template Id for the "plan" inside the organization
          via: "UI"
        },
        relationships: {
          workspace: {
            data: {
              type: "workspace",
              id: ${workspaceId}
            },
          },
        },
      },
    }
```

You can find some documentation about templates [here](https://docs.terrakube.io/user-guide/organizations/templates).

I hope this can help.
